api.go

```diff
@@ -0,0 +1,209 @@
+	}
+
+	gid, err := db.z.nextUID()
+	if err != nil {
+		return 0, object, err
+	}
+
+	dms := make([]*dql.Mutation, 0)
+	sch := &schema.ParsedSchema{}
+	err = generateSetDqlMutationsAndSchema[T](ctx, n, object, gid, &dms, sch)
+	if err != nil {
+		return 0, object, err
+	}
+
+	err = n.alterSchemaWithParsed(ctx, sch)
+	if err != nil {
+		return 0, object, err
+	}
+
+	err = applyDqlMutations(ctx, db, dms)
+	if err != nil {
+		return 0, object, err
+	}
+
+	return getByGid[T](ctx, n, gid)
+}
+
+func Upsert[T any](db *DB, object T, ns ...uint64) (uint64, T, bool, error) {
+
+	var wasFound bool
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+	if len(ns) > 1 {
+		return 0, object, false, fmt.Errorf("only one namespace is allowed")
+	}
+
+	ctx, n, err := getDefaultNamespace(db, ns...)
+	if err != nil {
+		return 0, object, false, err
+	}
+
+	gid, cfKeyValue, err := structreflect.GetUniqueConstraint[T](object)
+	if err != nil {
+		return 0, object, false, err
+	}
+	var cf *ConstrainedField
+	if cfKeyValue != nil {
+		cf = &ConstrainedField{
+			Key:   cfKeyValue.Key(),
+			Value: cfKeyValue.Value(),
+		}
+	}
+
+	dms := make([]*dql.Mutation, 0)
+	sch := &schema.ParsedSchema{}
+	err = generateSetDqlMutationsAndSchema[T](ctx, n, object, gid, &dms, sch)
+	if err != nil {
+		return 0, object, false, err
+	}
+
+	err = n.alterSchemaWithParsed(ctx, sch)
+	if err != nil {
+		return 0, object, false, err
+	}
+
+	if gid != 0 || cf != nil {
+		gid, err = getExistingObject[T](ctx, n, gid, cf, object)
+		if err != nil && err != apiutils.ErrNoObjFound {
+			return 0, object, false, err
+		}
+		wasFound = err == nil
+	}
+
+	if gid == 0 {
+		gid, err = db.z.nextUID()
+		if err != nil {
+			return 0, object, false, err
+		}
+	}
+
+	dms = make([]*dql.Mutation, 0)
+	err = generateSetDqlMutationsAndSchema[T](ctx, n, object, gid, &dms, sch)
+	if err != nil {
+		return 0, object, false, err
+	}
+
+	err = applyDqlMutations(ctx, db, dms)
+	if err != nil {
+		return 0, object, false, err
+	}
+
+	gid, object, err = getByGid[T](ctx, n, gid)
+	if err != nil {
+		return 0, object, false, err
+	}
+
+	return gid, object, wasFound, nil
+}
+
+func Get[T any, R UniqueField](db *DB, uniqueField R, ns ...uint64) (uint64, T, error) {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+	var obj T
+	if len(ns) > 1 {
+		return 0, obj, fmt.Errorf("only one namespace is allowed")
+	}
+	ctx, n, err := getDefaultNamespace(db, ns...)
+	if err != nil {
+		return 0, obj, err
+	}
+	if uid, ok := any(uniqueField).(uint64); ok {
+		return getByGid[T](ctx, n, uid)
+	}
+
+	if cf, ok := any(uniqueField).(ConstrainedField); ok {
+		return getByConstrainedField[T](ctx, n, cf)
+	}
+
+	return 0, obj, fmt.Errorf("invalid unique field type")
+}
+
+func Query[T any](db *DB, queryParams QueryParams, ns ...uint64) ([]uint64, []T, error) {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+	if len(ns) > 1 {
+		return nil, nil, fmt.Errorf("only one namespace is allowed")
+	}
+	ctx, n, err := getDefaultNamespace(db, ns...)
```

Comment:
Here and in the rest of the API methods, we are creating the `ctx` context object ourselves inside the `getDefaultNamespace` function.  Instead, we should be allowing the user to pass context in as a parameter (usually the first parameter).

For example, say the user wants the query to timeout if it takes to long to run.  The normal way would be to use `context.WithTimeout` and pass that context in to the query function.  But we've not given them any way to do that.

Reply:
... Also, it's unclear why there is `Query` as a global function that doesn't take a context but takes a db and a namespace, and `db.Query` which _does_ take a context, but not a namespace.

IMHO, There should just be one way to query, not two.